src/lib/operators/aggregate_hash.cpp

```diff
@@ -27,43 +27,94 @@
 namespace {
 using namespace opossum;  // NOLINT
 
-// Given an AggregateKey key, and a RowId row_id where this AggregateKey was encountered, this first checks if the
-// AggregateKey was seen before. If not, a new aggregate result is inserted into results and connected to the row id.
-// This is important so that we can reconstruct the original values later. In any case, a reference to the result is
-// returned so that result information, such as the aggregate's count or sum, can be modified by the caller.
-template <typename ResultIds, typename Results, typename AggregateKey>
-typename Results::reference get_or_add_result(ResultIds& result_ids, Results& results, const AggregateKey& key,
-                                              const RowID& row_id) {
-  // Get the result id for the current key or add it to the id map
+// For a given AggregateKey key, and a RowId row_id where this AggregateKey was encountered, this returns a reference
+// to the AggregateResult for the group defined by the AggregateKey. If GroupBy is used, this includes a lookup in the
+// result_ids hashmap, which stores the AggregateKey->Results[i] mapping.
+//
+// If the operator calculates multiple aggregate functions, we only need to perform this lookup as part of the first
+// aggregate function. By setting CacheResultIds to true_type, we can store the result of the lookup in the
+// AggregateKey. Following aggregate functions can then retrieve the index from the AggregateKey.
+template <typename CacheResultIds, typename ResultIds, typename Results, typename AggregateKey>
+typename Results::reference get_or_add_result(CacheResultIds, ResultIds& result_ids, Results& results,
+                                              AggregateKey& key, const RowID& row_id) {
   if constexpr (std::is_same_v<AggregateKey, EmptyAggregateKey>) {
+    // The aggregation uses no GROUP BY. To avoid special handling, get_or_add_result is still called, however, we
+    // always return the same result.
     if (results.empty()) {
       results.emplace_back();
       results[0].row_id = row_id;
     }
     return results[0];
   } else {
-    auto it = result_ids.find(key);
-    if (it != result_ids.end()) return results[it->second];
+    // As described above, we may store the index into the results vector in the AggregateKey. As the AggregateKey may
+    // contain multiple entries, we always use its first entry and store a (non-owning, raw) pointer to that entry in
+    // first_key_entry.
+    AggregateKeyEntry* first_key_entry;
+    if constexpr (std::is_same_v<AggregateKey, AggregateKeyEntry>) {
+      first_key_entry = &key;
+    } else {
+      first_key_entry = &key[0];
+    }
 
-    auto result_id = results.size();
+    // If we store the result of the hashmap lookup (i.e., the index into results) in the AggregateKeyEntry, we do this
+    // by storing the index in the lower 63 bits of first_key_entry and setting the most significant bit to 1 as a
```

Comment:
Remember that x86 really likes things to be properly aligned. If we add a struct with a bool, it will be twice as big as the current solution, greatly f'ing up the cache behavior. We could try to pack things, but that will not improve the performance either.

Reply:
Hmm, it's probably a pain to measure that, right?
I know that we cannot nicely pack both values, but I am still not super happy with the current solutions. Feels like looking into MonetDB. Uargh ...

Reply:
This gets even worse. We would have to change the type of the `AggregateKey` for something that is limited only to `get_or_add_result`. That goes through half of the operator code. Just remembering how much we saved in the past by shaving off a couple of bytes here and there makes me somewhat convinced that this would be slower. Unless you insist on it, I wouldn't spend time on it.

Reply:
Nope, but maybe add the reasoning as a comment to make people aware why we are doing that here.